test/common/http/http1/codec_impl_test.cc

```diff
@@ -1225,6 +1225,23 @@ TEST_F(Http1ClientConnectionImplTest, UpgradeResponse) {
   codec_->dispatch(websocket_payload);
 }
 
+TEST_F(Http1ClientConnectionImplTest, UpgradeResponseWithNoBody) {
+  initialize();
+
+  InSequence s;
+
+  NiceMock<Http::MockStreamDecoder> response_decoder;
+  Http::StreamEncoder& request_encoder = codec_->newStream(response_decoder);
+  TestHeaderMapImpl headers{{":method", "HEAD"}, {":path", "/"}, {":authority", "host"}};
+  request_encoder.encodeHeaders(headers, true);
+
+  // Send upgrade headers. Make sure we avoid the deferred_end_stream_headers_ optimization.
+  EXPECT_CALL(response_decoder, decodeHeaders_(_, false));
+  Buffer::OwnedImpl response(
+      "HTTP/1.1 200 OK\r\nConnection: upgrade\r\nUpgrade: websocket\r\n\r\n");
+  codec_->dispatch(response);
```

Comment:
Should we allow upgrade with HEAD or just reset the connection?  I'm inclined to think that's pretty shady and we  may want to send an error response
@mattklein123  @PiotrSikora 

If we do allow HEAD and upgrade, I'm not sure if we'd want to perform the upgrade (which I think this change does) since HEAD is basically supposed to inform one of the response headers one would get if one did a GET.  I think we'd really need to special case it.  I'd definitely want an integration test in that case.

Reply:
IMO probably better to just 400 it, but will defer to what you and @PiotrSikora think.

Reply:
`HEAD` is legit, e.g. HTTP/1.1-to-HTTP/2 upgrade where the first request is `HEAD`, i.e.:
```
HEAD / HTTP/1.1
Host: example.com
Connection: Upgrade, HTTP2-Settings
Upgrade: h2c
HTTP2-Settings: ...
```
I'm aware that Envoy doesn't support HTTP/1.1-to-HTTP/2 upgrade just yet, but we're supposed to respond with HTTP/1.1 response to such request.

Reply:
In response the `upgrade: h2c` case, Envoy won't crash like it does here, since we'll bypass upgrade mode. (And only in upgrade mode do we assert that there's no deferred end stream headers).

>  but we're supposed to send regular HTTP/1.1 response to such request.

I take this to mean that we should allow `HEAD` and upgrade, but skip the upgrade?

Reply:
Correct, upgrade is always optional (unless enforced by the client with `OPTIONS` method).

Reply:
Is it too unclear that the base sets `handling_upgrade_` but removes this when we find the response is following a `HEAD` request? I'm not sure if there's a way to test for being in 'a response to a `HEAD` request scenario' when we set `handling_upgrade_` in `onHeadersCompleteBase` (rather than waiting till we get to the `ClientConnectionImpl`).

Reply:
Erm, sorry if what I said was confusing... I meant that the upgrade request will be ignored and that the upstream will respond with HTTP/1.1 response, because we don't support HTTP/1.1-to-HTTP/2 upgrades in general. This is not unique to the `HEAD` method.

`Upgrade: h2c` header should be already stripped in [`ConnectionImpl::onHeadersCompleteBase()`](https://github.com/envoyproxy/envoy/blob/1a7b32de759939f82af2cf7e83471aff5d57c93d/source/common/http/http1/codec_impl.cc#L512), so I'm confused why it would be still available in `ClientConnectionImpl::onHeadersComplete()` and why `handling_upgrade_` isn't `false`?

Reply:
Hmmm. We also have `websocket` upgrades, though. So the current codepath does a check for if we have upgrade headers. And if we have`h2c`, we strip the relevant headers. Otherwise, we set `handling_upgrade_` to true and upgrade for websocket.

Reply:
Bump on this? The upgrade behavior is:
- if `h2c`, strip upgrade headers and ignore the upgrade (already implemented, unchanged)
- if upgrade headers are present with anything else, handle as a `websocket` upgrade
- skip the upgrade if it's a response to a `HEAD` and remove upgrade headers

There's an integration test for this. Piotr, does that make sense, or should I just 400 it?

Reply:
Yeah, if we're stripping upgrades for h2c I would mildly prefer disallowing over allowing-and-altering.  Again I'm Ok going with either, especially as we're not catching this on the request path so it'd be 5xx response rather than an early 4xx.  Happy to defer to Piotr and Matt on it though

Reply:
Erroring out HEAD requests with upgrade headers seems like an OK way to avoid this crash condition.

What upgrade types does Envoy support?  I assume only websocket.

Reply:
That's correct, websocket is the only one supported. h2c upgrades are stripped.

I'll re-open and make the 400 fix on Monday :)